cpp/src/arrow/engine/substrait/relation_internal.cc

```diff
@@ -683,6 +785,52 @@ Result<DeclarationInfo> FromProto(const substrait::Rel& rel, const ExtensionSet&
                          std::move(aggregate_schema));
     }
 
+    case substrait::Rel::RelTypeCase::kExtensionLeaf:
+    case substrait::Rel::RelTypeCase::kExtensionSingle:
+    case substrait::Rel::RelTypeCase::kExtensionMulti: {
+      std::vector<DeclarationInfo> ext_rel_inputs;
+      ARROW_ASSIGN_OR_RAISE(
+          auto ext_rel_info,
+          GetExtensionRelationInfo(rel, ext_set, conversion_options, &ext_rel_inputs));
+      const auto& ext_decl_info = ext_rel_info.decl_info;
+      auto ext_common_opt = GetExtensionRelCommon(rel);
+      bool has_emit = ext_common_opt && ext_common_opt->emit_kind_case() ==
+                                            substrait::RelCommon::EmitKindCase::kEmit;
+      if (!ext_rel_info.field_output_indices) {
+        if (!has_emit) {
+          return ProcessEmitProject(ext_common_opt, ext_decl_info,
+                                    ext_decl_info.output_schema);
+        }
```

Comment:
```suggestion
        if (!has_emit) {
          return ext_decl_info;
        }
```

Given that we know there is no emit that is all `ProcessEmitProject` is going to do anyways.  Then we can get rid of the change you made introducing `ProcessEmitProject` (e.g. creating the helper function) because we don't need it anymore.

Reply:
Fixed in the recent commit. However, `ProcessEmitProject` is invoked from another place, so I kept it.